pyiron_base/job/template.py

```diff
@@ -24,17 +26,41 @@ class TemplateJob(GenericJob, HasStorage):
     def __init__(self, project, job_name):
         GenericJob.__init__(self, project, job_name)
         HasStorage.__init__(self)
-        self.storage.create_group("input")
-        self.storage.create_group("output")
+        self.storage.input = self._input_class(table_name='input')
+        self.storage.output = self._output_class(table_name='output')
 
     @property
-    def input(self):
+    def _input_class(self) -> Type[DataContainer]:
+        """Children can overwrite this method to return some other child of DataContainer for custom behaviour"""
+        return DataContainer
+
+    @property
+    def _output_class(self) -> Type[DataContainer]:
+        """Children can overwrite this method to return some other child of DataContainer for custom behaviour"""
+        return DataContainer
+
+    @property
+    def input(self) -> Type[DataContainer]:
         return self.storage.input
 
+    @input.setter
+    def input(self, new_input):
+        raise AttributeError(
+            "Input cannot be overwritten; to get custom output behaviour, override the `_input_class` property to "
+            "return your own custom class (which inherits from `DataContainer`)."
```

Comment:
Nice! And I guess I'd even give a simple example:

```python
Example:

>>> class MyInput(DataContainer):
>>>     def __init__(self, init=None, table_name=None, lazy=False, wrap_blacklist=()):
>>>         super().__init__(init=init, table_name=table_name, lazy=lazy, wrap_blacklist=wrap_blacklist)
>>>         self.foo = 5
>>>
>>>
>>> job.input._input_class = MyInput()
```

We can maybe also mention `from pyiron_base.generic.datacontainer import DataContainer`

Reply:
An example is a stellar idea, but I think we'll need to rework it based on my serialization comment. Adding an example leans into your high-level comment about really helping developers, which I love!

Reply:
aaaah nice now with the example I actually understood how it really works XDXDXDXDXD